tests/test_introduction_to_classes.py

```diff
@@ -0,0 +1,197 @@
+    True,
+    False,
+    0,
+    -1,
+    ""
+])
+@pytest.mark.parametrize("test_animal_name_neg", [
+    None,
+    True,
+    False,
+    0
+])
+def test_description_negative(test_animal_name_neg, test_animal_age_neg):
+    try:
+        hippo = Animal(test_animal_name_neg, test_animal_age_neg)
+        hippo.description()
+        raise Exception("Expected an exception from test_description_negative")
+    except AssertionError as e:
+        assert str(e) == "Error type, please enter a valid value age or name"
+
+
+@pytest.mark.parametrize("input_val", [
+    1,
+    12,
+    123,
+])
+def test_display_cart(input_val):
+    customer_test = Customer(input_val)
+    assert customer_test.display_cart() == input_val
+
+
+@pytest.mark.parametrize("input_neg", [
+    None,
+    True,
+    False,
+    0,
+    -1,
+    "",
+    "A"
+])
+def test_display_cart_negative(input_neg):
+    try:
+        customer_test = Customer(input_neg)
+        customer_test.display_cart()
+        raise Exception("Expected an exception from test_display_cart_negative")
+    except AssertionError as e:
+        assert str(e) == "Error type, please enter a valid value"
+
+
+@pytest.mark.parametrize("input_val", [
+    1,
+    12,
+    123
+])
+def test_display_order_history(input_val):
+    order_test = ReturningCustomer(input_val)
+    assert order_test.display_order_history() == "I'm a string that stands in for your order history!"
+
+
+@pytest.mark.parametrize("input_neg", [
+    None,
+    True,
+    False,
+    0,
+    -1,
+    "",
+    "A"
+])
+def test_display_cart_negative(input_neg):
+    try:
+        order_test = ReturningCustomer(input_neg)
+        order_test.display_order_history()
+        raise Exception("Expected an exception from display_order_history")
+    except AssertionError as e:
+        assert str(e) == "Error type, please enter a valid value"
+
+
+@pytest.mark.parametrize("test_angle1, test_angle2, test_angle3, expected", [
+    (0, 0, 180, True),
+    (1, 179, 0, True),
+    (1, 1, 178, True),
+    (59, 62, 59, True),
+    (-180, 180, 180, True),
+    (180, 1, 0, False)
+])
+def test_check_angles(test_angle1, test_angle2, test_angle3, expected):
+    check_angles = Triangle(test_angle1, test_angle2, test_angle3)
+    assert check_angles.check_angles() == expected
+
+
+@pytest.mark.parametrize("test_angle1", [
+    None,
+    True,
+    False,
+    "",
+    "A"
+])
+@pytest.mark.parametrize("test_angle2", [
+    None,
+    True,
+    False,
+    "",
+    "A"
+])
+@pytest.mark.parametrize("test_angle3", [
+    None,
+    True,
+    False,
+    "",
+    "A"
+])
+def test_check_angles_neg(test_angle1, test_angle2, test_angle3):
+    try:
+        check_angles = Triangle(test_angle1, test_angle2, test_angle3)
+        check_angles.check_angles()
+        raise Exception("Expected an exception from test_display_cart_negative")
+    except AssertionError as e:
+        assert str(e) == "Error type, please enter a valid value"
+
+
+
+
+
+
+
+
+
+
```

Comment:
minor style: too many empty lines at the end of the file. Should only 1 empty line.